src/haz3lcore/statics/Statics.re

```diff
@@ -636,3 +636,82 @@ let collect_errors = (map: Map.t): list((Id.t, Info.error)) =>
     map,
     [],
   );
+
+/*
+ * Check whether the argument of a function under the new annotation is of correct format,
+ * and output a list of input types
+ */
+let caf_argtrans =
+    (args: list(UPat.t)): option((list(UPat.t), list(UTyp.t))) =>
+  List.fold_right(
+    (item: UPat.t, partial: option((list(UPat.t), list(UTyp.t)))) =>
+      switch (partial) {
+      | Some((lst_var, lst_typ)) =>
+        switch (item.term) {
+        | TypeAnn({ids: inner_ids, term: Var(var_name)}, t) =>
+          Some((
+            [{ids: inner_ids, term: Var(var_name)}, ...lst_var],
+            [t, ...lst_typ],
+          ))
+        | Var(_) =>
+          Some((
+            [item, ...lst_var],
+            [{ids: item.ids, term: EmptyHole}, ...lst_typ],
+          ))
+        | _ => None
+        }
+      | _ => None
+      },
+    args,
+    Some(([], [])),
+  );
+
+/*
+ * Check whether a particular let binding is an annotated function under the new syntax.
+ * It is parsed as a constructor syntax.
+ * I am not sure what is the top-level pattern; the only thing similar seems to be UPat.Ap.
+ */
+let check_annotated_function =
+    (pat: UPat.t): option((Var.t, UPat.t, UTyp.t)) =>
+  switch (pat.term) {
+  | TypeAnn({ids: inner_ids, term: Ap(var, args)}, ret_type) =>
+    // Check whether all arguments are parameters with annotations
+    switch (var.term, args.term) {
+    | (Var(func_name), Tuple(ps)) =>
+      switch (caf_argtrans(ps)) {
+      | None => None
+      | Some((params, in_type)) =>
+        Some((
+          func_name,
+          {ids: args.ids, term: Tuple(params)},
```

Comment:
you don't need to reconstruct this, since this record is already the `args` variable (and similar below)